docs/sc-memory/api/cpp/extended/permissions_api.md

```diff
@@ -0,0 +1,809 @@
+
+```cpp
+ScMemoryContext context;
+ScAddr const & guestUserAddr = context.GetUser();
+// This user should belongs 
+```
+
+## **How does the sc-machine identifies users?**
+
+You can identify user. User identification refers to the process of identifying a user on ostis-system, i.e., that the specified guest user is some user that is on the system.
+
+It is useful when your ostis-system implements functionality to work with guest users. The sc-machine provides users identification. You can identify some guest user as user that had been already registered in the system. Users identification allows you to implement your own agent to register/authorize users by password, token or something else.
+
+To identify user in the sc-memory, you should create connection between guest user and identified user.
+
+```scs
+..some_guest_user => nrel_identified_user: ..some_user;;
+```
+
+After that, all contexts that had specified guest user will have specified identified user. 
+
+To unidentify identified user, you can create reverse connection between specified guest user and specified identified user.
+
+## **How does the sc-machine authenticate users?**
+
+Users can be known in advance and then guest users and their identification is unnecessary, but all users should be authenticated in sc-memory. User authentication involves the process of verifying the authenticity of identified user and storing information about that user in sc-memory.
+
+To authenticate user in sc-memory, you should create connection between `concept_authentication_request` and user:
+
+```scs
+concept_authentication_request -> ..some_user;;
+```
+
+You should do it once, no more. After the sc-machine authenticates this user, it means that connection between `concept_authenticated_user` and user has been created and all contexts with this user have been authenticated.
+
+```scs
+concept_authenticated_user ~> ..some_user;;
+```
+
+If you remove this sc-arc between `concept_authenticated_user` and the user, then the user and all contexts with user will be also unauthenticated.
+
+## **Examples of using user identification and authentication**
+
+### **Example of user identification**
+
+You should identify guest user before authentication of user that will be identified. 
+
+```cpp
+ScMemoryContext context; // Some context with guest user.
+ScAddr const & guestUserAddr = context.GetUser();
+
+// Identify guest user.
+// Find a user, which was identified. You can provide your own logic of user
+// identification before.
+ScAddr const & userAddr1 = context.HelperFindBySystemIdtf("user_1");
+
+ScAddr const & arcAddr = ScMemory::ms_globalContext->CreateEdge(
+  ScType::EdgeDCommonConst, guestUserAddr, userAddr1);
+ScMemory::ms_globalContext->CreateEdge(
+  ScType::EdgeAccessConstPosPerm, ScKeynodes::nrel_identified_user, arcAddr);
+// Only `ScMemory::ms_globalContext` can identify users.
```

Comment:
Should we wait finish of sc-memory callback ?

Reply:
No, because there is no pragmatics in waiting for user identification to complete. After identification, the user must be authenticated

Reply:
let's discuss

Reply:
Fixed